modules/reindex/src/main/java/org/elasticsearch/index/reindex/ScrollableHitSource.java

```diff
@@ -0,0 +1,364 @@
+     */
+    public static class SearchFailure implements Writeable, ToXContent {
+        private final Throwable reason;
+        @Nullable
+        private final String index;
+        @Nullable
+        private final Integer shardId;
+        @Nullable
+        private final String nodeId;
+
+        public SearchFailure(Throwable reason, @Nullable String index, @Nullable Integer shardId, @Nullable String nodeId) {
+            this.index = index;
+            this.shardId = shardId;
+            this.reason = requireNonNull(reason, "reason cannot be null");
+            this.nodeId = nodeId;
+        }
+
+        /**
+         * Build a search failure that doesn't have shard information available.
+         */
+        public SearchFailure(Throwable reason) {
+            this(reason, null, null, null);
+        }
+
+        public SearchFailure(StreamInput in) throws IOException {
+            index = in.readOptionalString();
+            shardId = in.readOptionalVInt();
+            reason = in.readThrowable();
+            nodeId = in.readOptionalString();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeOptionalString(index);
+            out.writeOptionalVInt(shardId);
+            out.writeThrowable(reason);
+            out.writeOptionalString(nodeId);
+        }
+
+        public String getIndex() {
+            return index;
+        }
+
+        public Integer getShardId() {
+            return shardId;
+        }
+
+        public Throwable getReason() {
+            return reason;
+        }
+
+        @Nullable
+        public String getNodeId() {
+            return nodeId;
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            if (index != null) {
+                builder.field("index", index);
+            }
+            if (shardId != null) {
+                builder.field("shard", shardId);
+            }
+            if (nodeId != null) {
+                builder.field("node", nodeId);
+            }
+            builder.field("reason");
+            {
+                builder.startObject();
+                ElasticsearchException.toXContent(builder, params, reason);
+                builder.endObject();
+            }
+            builder.endObject();
+            return builder;
+        }
+
+        @Override
+        public String toString() {
+            return Strings.toString(this);
+        }
+
+        public String locationInfo() {
+            StringBuilder b = new StringBuilder();
+            if (index != null) {
+                b.append("index=").append(index);
+            }
+            if (shardId != null) {
+                b.append("shard=").append(shardId);
+            }
+            if (nodeId != null) {
+                b.append("node=").append(nodeId);
```

Comment:
this is going to look a little strange as `index=foobarbasshard=0node=thing`, can we split it apart?